starlette/middleware/http.py

```diff
@@ -0,0 +1,94 @@
+from contextlib import aclosing
+from functools import partial
+from typing import AsyncGenerator, Callable, Optional
+
+from ..datastructures import MutableHeaders
+from ..responses import Response
+from ..types import ASGIApp, Message, Receive, Scope, Send
+
+HTTPDispatchFlow = AsyncGenerator[Optional[Response], Response]
+
+
+class HTTPMiddleware:
+    def __init__(
+        self,
+        app: ASGIApp,
+        dispatch_func: Optional[Callable[[Scope], HTTPDispatchFlow]] = None,
+    ) -> None:
+        self.app = app
+        self.dispatch_func = self.dispatch if dispatch_func is None else dispatch_func
+
+    def dispatch(self, scope: Scope) -> HTTPDispatchFlow:
+        raise NotImplementedError  # pragma: no cover
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        async with aclosing(self.dispatch(scope)) as flow:
+            # Kick the flow until the first `yield`.
+            # Might respond early before we call into the app.
+            early_response = await flow.__anext__()
+
+            if early_response is not None:
+                await early_response(scope, receive, send)
+                return
+
+            response_started = set[bool]()
+
+            wrapped_send = partial(
+                self._send,
+                flow=flow,
+                response_started=response_started,
+                send=send,
+            )
```

Comment:
Can we just make this a closure?

Reply:
No strong sentiments on that, I tend to prefer the partial method style as it keeps the main body of the middleware free from clutter, but the result here with everything inline is acceptable too.

Reply:
I seem to remember someone in another issue complaining about us changing an internal method. If it's not intended for subclasses to use, can we name mangle it (`__` I think)?

Reply:
You might be looking at the outdated diff -- I switched to the closure style, so no more `_send()` method. :+1: